### PR TITLE
User management RFC

### PR DESCRIPTION
**Pull request by kelvich**
[Rendered](https://github.com/neondatabase/neon/blob/user_mgmt_rfc/docs/rfcs/017-user-management.md)

### CODE REVIEW

File: docs/rfcs/017-user-management.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+* we want to be able to authenticate users in proxy against the console without Postgres involvement. Otherwise,
+malicious brute force attempts will wake up Postgres (expensive) and may exhaust the Postgres connection pool (deny of service).
+* it is handy when we can render console UI without waking up compute (e.g., show database list)
+
+Storing the same information in two systems is a form of replication. And in the current scheme
+the console is primary, and Postgres catalog is a replica.
+
+This RFC proposes to address problems 1. and 2. by making Postgres a source of truth for roles/databases and
+only caching this info in Postgres. So using the replication analogy, now the Postgres catalog will be primary, and
```
Comment (ololobus) [inline docs/rfcs/017-user-management.md]: caching this info in Postgres -> caching this info in **the console**

Comment (stepashka) [inline docs/rfcs/017-user-management.md:20]:
```suggestion
only caching this info in the console. So using the replication analogy, now the Postgres catalog will be primary, and
```

File: docs/rfcs/017-user-management.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+the console will be a replica. Problem 3 is a bit different and could be addressed by ditching the web_access
+user and using, e.g., JWT auth for the @username user so that we do not introduce a new user (JWT is needed
+since we don't know users password).
+
+This RFC doesn't talk about giving root access to the database, which is blocked by a secure runtime setup.
+
+## Overview
+
+* Add `/tenant/$tenant/branch/$branch/refresh_catalog` endpoint to console management API which asks `/get_catalog` and updates cached roles/databases info.
```
Comment (ololobus) [inline docs/rfcs/017-user-management.md:29]: asks `/get_catalog` from the compute_ctl http API, I suppose

File: docs/rfcs/017-user-management.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+* Whenever user edits list of databases or users postgres signals `compute_ctl` to call `/<...>/refresh_catalog` in the console
```
Comment (ololobus) [inline docs/rfcs/017-user-management.md:30]: I think that Postgres can do a direct call to the control plane API. Not sure how is it difficult to do an empty POST HTTP request from the C, though

File: docs/rfcs/017-user-management.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+* Add password strenght check in our extension
+
+## Postgres behavior
+
+Default user role (@username) should have `CREATE ROLE`, `CREATE DB`, and `BYPASSRLS` privileges. We expose Postgres port
+to the open internet, so we need to check passwords strength. We can use the `passwordcheck` extension or do the same
+from our extension.
+
+Whenever a user edits a list of databases or users, Postgres sends SIGHUP to `compute_ctl`. `compute_ctl` should write PID to `compute_ctl.pid` file.
```
Comment (ololobus) [inline docs/rfcs/017-user-management.md:39]: As mentioned above maybe we can just call the http api directly from the extension? We don't transfer any data, it's just a signal call, so should be safe and +- easy

File: docs/rfcs/017-user-management.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+* we, anyway, will need console originated call to compute's `/get_catalog` after historical branch creation
+* If an intruder gains access to some other `/tenant/$tenant/.../refresh_catalog` he won't be able to change the roles list and will just force an unnecessary reload.
+
+`/refresh_catalog` returns HTTP 200 OK on success.
+
+We should have a button in the admin UI to manually force `/refresh_catalog` in case of data desync.
+
+# Scalability
+
+On my laptop, I can create 4200 roles per second. That corresponds to 363 million roles per day. So both `/get_catalog` can become expensive, and our roles database can snowball. While we can address `/get_catalog` size by catching only the latest changes (e.g., maintain the audit table and drain it by the console), it is still not nice that a single tenant can blow up a multi-tenant console database. I would instead propose to limit the number of databases and roles by some big number like 1000 and bump this limit if somebody asks for it with a legit use case.
```
Comment (ololobus) [inline docs/rfcs/017-user-management.md:73]: Interesting, didn't thought about this. Some large limit sounds reasonable to me

**Review by ololobus — APPROVED** [review]
Thanks for writing this. It almost exactly matches my understanding of how it could be designed. Few more thoughts:
- taking into account the planned console split into web part and control-plane, let's mention control-plane explicitly whenever it applies (in most places) to avoid possible misreading.
- you don't mention how users and dbs can be created from the console UI. I still think that it's nice feature to have for not experienced-postgres-hackers. This is how I imagined it:
   * once we have this JWT auth, we can use it for `cloud_admin`
   * so we again create an `apply_config` operation; or, if compute is shutdown, we create a chain `start_compute -> apply_config`
   * operation is needed for persistence and idempotency of the change
   * then `apply_config` just uses a libpq protcol to create role / db, do rename or whatever we want

Thoughts?

**Review by hlinnaka — APPROVED** [review]
Sounds good, definitely a step in good direction.

In the long run, I've got a feeling that the console should not try to maintain a copy of all the users and databases. If a user wants to see the list of databases, wake up the compute and fetch the list. We can cache the list in the console for a short while for performance. Similarly for authentication, I think a small cache of users would suffice, instead of trying to mirror everything in the console.

In time-travel situations, I don't think we can store a copy of all users and databases *at all points in time*. Some kind of lazy loading will be needed in that scenario anyway.

**Comment by kelvich** [comment]
> Similarly for authentication, I think a small cache of users would suffice, instead of trying to mirror everything in the console.

Hmm. So, if Alice tries to brute force usernames on an endpoint, compute will wake, and Alice can keep it always awake since a non-existent user is always a cache miss. Moreover, if she can somehow get a big enough list of our endpoint addresses, she can try to simultaneously wake them, exhausting all our pre-created k8s nodes so that a new legit compute won't start. That was my main rationale behind the decision to store all roles on the console (probably, I should've written this in RFC itself).

Agree that storing databases in console is optional, and we can wake compute and query it instead to render pages.

> In time-travel situations, I don't think we can store a copy of all users and databases at all points in time. Some kind of lazy loading will be needed in that scenario anyway.

I was thinking of the model where the console calls `/get_catalog` right after branch creation, and then the proxy can authenticate connections against it. So we will store all roles current_number_of_branches times.

Tricky question here is what to do when we want to query historical data by using only connection options (e.g. `postgres://<...>?asoftime=12341314`) -- there is no branch yet, hence no cached info yet. I can think of three possible behaviors:
1) Let any connection spin up a compute and do auth when it will be running. If we will limit free tier users with N simultaneous computes, this behavior will make wake-up-call attack N times worse.
2) Do not allow branch/ro-compute creation via connection options, require API call for that. In this model, we can still get away with oneliner for historical connection, e.g.
```
> psql `curl -H 'Auth...' api.neon.tech/projects/rough-queen-833418/get_temp_branch?time=1234`
```
given that `/get_temp_branch` will return a connection string. By that time, compute will be already started, and roles will be available.
3) Combination of 1) and 2) -- try authorizing against the current state of roles. This will work if your password on the branch is the same. If it is not, then use API. This will work in most cases, but I don't like that there is quite a lot of implicit behavior.
4) Actually store all the role changes in the console

I actually think that option 2) could be even nicer/clearer compared to `?options=` way. There is less magic, and you know immediately what is happening. Also, `%3D` in options is quite annoying.

**Comment by ololobus** [comment]
> Tricky question here is what to do when we want to query historical data by using only connection options (e.g. postgres://<...>?asoftime=12341314) -- there is no branch yet, hence no cached info yet. I can think of several possible behaviors:

I also thought that we can have an explicit checkbox on the endpoint that it allows time-travel, which automatically implies that it's read-only. After that we can implicitly create a new branch, spin up the compute on it, fetch users, authenticate, use and delete branch when compute is suspended.

We can also set a limit, i.e. if you have a connection to this endpoint @LSN1 and try to connect @LSN2, we will restart the compute to work on the LSN2 and first connection will be dropped.

**Comment by hlinnaka** [comment]
> > Similarly for authentication, I think a small cache of users would suffice, instead of trying to mirror everything in the console.
> 
> Hmm. So, if Alice tries to brute force usernames on an endpoint, compute will wake, and Alice can keep it always awake since a non-existent user is always a cache miss. Moreover, if she can somehow get a big enough list of our endpoint addresses, she can try to simultaneously wake them, exhausting all our pre-created k8s nodes so that a new legit compute won't start. That was my main rationale behind the decision to store all roles on the console (probably, I should've written this in RFC itself).

Yeah, that would be bad. I don't have a great solution in mind. Caching full list of users is one way, I agree. Another might be some generic DDoS protection, blocking too-frequent attempts from certain IP addresses.

P.S [Alice](https://en.wikipedia.org/wiki/Alice_and_Bob) would never do such a thing. Sounds more like Mallory ;-)

**Comment by LizardWizzard** [comment]
As discussed moving to draft.

**Comment by vadim2404** [comment]
closed in favor of #3838

**Comment by vadim2404** [comment]
Implemented in accordance with that PR. It will be merged

**Comment by Ranguna** [comment]
@vadim2404 which PR will be merged?

**Comment by vadim2404** [comment]
yes, my previous message sounds a bit weird :D sorry

#3838 will be merged

**Comment by Ranguna** [comment]
Ah, so it will be "unclosed" and merged?

Nice, thanks!

**Comment by Ranguna** [comment]
Hmmm wait a second, if it's going to be merged in #3848, why did you close it today together with this one ?

Maybe a missclick?

**Comment by vadim2404** [comment]
@Ranguna , thanks. Yes, you're right. #3838 should be merged. This one - must be closed.

**Comment by Ranguna** [comment]
Awesome thanks for the clarification!

**Pull request closed without merging**